platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/DefaultValueSourceProviderFactory.java

```diff
@@ -277,33 +286,31 @@ private LazilyObtainedValue(
             this.sourceType = sourceType;
             this.parametersType = parametersType;
             this.parameters = parameters;
+            this.value = calculatedValueFactory.create(Describables.of("ValueSource of type", sourceType), () -> {
+                    computationBroadcaster.getSource().beforeValueObtained();
+                    try {
+                        ValueSource<T, P> source = source();
+                        sourceRef.set(source);
+                        return source.obtain();
+                    } finally {
+                        computationBroadcaster.getSource().afterValueObtained();
+                    }
+                }
+            );
         }
 
         public boolean hasBeenObtained() {
-            return value != null;
+            return value.isFinalized();
         }
 
         public Try<@org.jetbrains.annotations.Nullable T> obtain() {
-            ValueSource<T, P> source;
-            // Return value from local to avoid nullability warnings when returning value from the field directly.
-            Try<T> obtained;
-            synchronized (this) {
-                Try<T> cached = value;
-                if (cached != null) {
-                    return cached;
-                }
-                computationBroadcaster.getSource().beforeValueObtained();
-                try {
-                    // TODO - add more information to exceptions
-                    // Fail fast when source can't be instantiated.
-                    source = source();
-                    value = obtained = Try.ofFailable(source::obtain);
-                } finally {
-                    computationBroadcaster.getSource().afterValueObtained();
-                }
+            value.finalizeIfNotAlready();
+            Try<@org.jetbrains.annotations.Nullable T> obtained = value.getValue();
+            ValueSource<T, P> obtainedFrom = sourceRef.getAndSet(null);
```

Comment:
> 🤔 Should the `sourceRef` cleaning be in a finally block just in case?

It doesn't hurt that much, and there's slight possibility of some `RuntimeException(InterruptedException)` popping up from `finalizeIfNotAlready`, so let's have that.